File: fetcher.py
import requests
from typing import List, Dict, Any
from models import Room, NotLoginError, DouyuAPIError
from config import API_URL


def fetch_douyu_live_status(cookies: Dict[str, str]) -> List[Room]:
    """
    Fetch Douyu live status from the follow list API.
    
    Args:
        cookies: Dictionary of cookies for authentication
        
    Returns:
        List of Room objects
        
    Raises:
        NotLoginError: If user is not logged in
        DouyuAPIError: For other API errors
    """
    cookie_str = "; ".join(f"{k}={v}" for k, v in cookies.items())
    headers = {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
        "Accept": "application/json, text/plain, */*",
        "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8",
        "Referer": "https://www.douyu.com/",
        "Cookie": cookie_str
    }

    try:
        response = requests.get(
            API_URL,
            headers=headers,
            timeout=30
        )
        response.raise_for_status()
        data = response.json()

        # Check for login status
        msg = data.get("msg") or ""
        error = data.get("error")
        if str(error) == "-1" and "未登" in msg:
            raise NotLoginError(msg)

        if str(error) != "0":
            raise DouyuAPIError(f"API error: {msg or data}")
        
        return parse_response(data)
        
    except requests.exceptions.RequestException as e:
        raise DouyuAPIError(f"Request failed: {e}")


def parse_response(data: Dict[str, Any]) -> List[Room]:
    """Parse Douyu API response into Room objects."""
    rooms = []
    room_list = data.get("data", {}).get("list", [])
    
    for room_data in room_list:
        room_id = str(room_data.get("room_id", ""))
        room_name = room_data.get("room_name", "")
        streamer_name = room_data.get("nickname", "")
        cover = room_data.get("room_src", "")
        avatar = room_data.get("avatar_small", "")
        show_status = room_data.get("show_status", 0)
        video_loop = room_data.get("videoLoop", 0)
        area_name = room_data.get("game_name", "")
        path = room_data.get("url", f"/{room_id}")
        
        # is_live: true if show_status == 1 AND videoLoop == 0
        is_live = (show_status == 1 and video_loop == 0)
        
        room = Room(
            room_id=f"dy_{room_id}",
            room_name=room_name,
            streamer_name=streamer_name,
            cover=cover,
            avatar=avatar,
            is_live=is_live,
            area_name=area_name,
            url=f"https://www.douyu.com{path}"
        )
        rooms.append(room)
    
    return rooms

File: fetcher.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sort"
	"strconv"
	"strings"
)

type httpDoer interface {
	Do(*http.Request) (*http.Response, error)
}

var douyuHTTPClient httpDoer = &http.Client{Timeout: douyuRequestTimeout}

type douyuAPIResponse struct {
	Message json.RawMessage `json:"msg"`
	Error   json.RawMessage `json:"error"`
	Data    struct {
		Rooms []map[string]json.RawMessage `json:"list"`
	} `json:"data"`
}

// fetchDouyuLiveStatus fetches and parses the authenticated Douyu follow list.
func fetchDouyuLiveStatus(ctx context.Context, cookies map[string]string) ([]Room, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, douyuAPIURL, nil)
	if err != nil {
		return nil, &DouyuAPIError{Message: fmt.Sprintf("Request failed: %v", err)}
	}

	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
	req.Header.Set("Accept", "application/json, text/plain, */*")
	req.Header.Set("Accept-Language", "zh-CN,zh;q=0.9,en;q=0.8")
	req.Header.Set("Referer", "https://www.douyu.com/")
	req.Header.Set("Cookie", cookieHeader(cookies))

	response, err := douyuHTTPClient.Do(req)
	if err != nil {
		if ctx.Err() != nil {
			return nil, ctx.Err()
		}
		return nil, &DouyuAPIError{Message: fmt.Sprintf("Request failed: %v", err)}
	}
	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, &DouyuAPIError{Message: fmt.Sprintf("Request failed: %v", err)}
	}
	if response.StatusCode < http.StatusOK || response.StatusCode >= http.StatusBadRequest {
		return nil, &DouyuAPIError{Message: fmt.Sprintf("Request failed: HTTP status %s", response.Status)}
	}

	var payload douyuAPIResponse
	if err := json.Unmarshal(body, &payload); err != nil {
		return nil, &DouyuAPIError{Message: fmt.Sprintf("Request failed: invalid JSON response: %v", err)}
	}

	message := rawText(payload.Message, "")
	errorCode := rawScalarString(payload.Error, "None")
	if errorCode == "-1" && strings.Contains(message, "未登") {
		return nil, &NotLoginError{Message: message}
	}
	if errorCode != "0" {
		detail := message
		if detail == "" {
			detail = strings.TrimSpace(string(body))
		}
		return nil, &DouyuAPIError{Message: fmt.Sprintf("API error: %s", detail)}
	}

	return parseResponse(payload), nil
}

func cookieHeader(cookies map[string]string) string {
	keys := make([]string, 0, len(cookies))
	for key := range cookies {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	parts := make([]string, 0, len(keys))
	for _, key := range keys {
		parts = append(parts, key+"="+cookies[key])
	}
	return strings.Join(parts, "; ")
}

// parseResponse converts a successful Douyu response into Room values.
func parseResponse(payload douyuAPIResponse) []Room {
	rooms := make([]Room, 0, len(payload.Data.Rooms))
	for _, roomData := range payload.Data.Rooms {
		roomID := rawScalarString(roomData["room_id"], "")
		pathRaw, hasPath := roomData["url"]
		path := "/" + roomID
		if hasPath {
			path = rawText(pathRaw, "")
		}

		showStatus := rawNumber(roomData["show_status"], 0)
		videoLoop := rawNumber(roomData["videoLoop"], 0)
		rooms = append(rooms, Room{
			RoomID:       "dy_" + roomID,
			RoomName:     rawText(roomData["room_name"], ""),
			StreamerName: rawText(roomData["nickname"], ""),
			Cover:        rawText(roomData["room_src"], ""),
			Avatar:       rawText(roomData["avatar_small"], ""),
			IsLive:       showStatus == 1 && videoLoop == 0,
			AreaName:     rawText(roomData["game_name"], ""),
			URL:          "https://www.douyu.com" + path,
			Platform:     "douyu",
		})
	}
	return rooms
}

func rawText(raw json.RawMessage, fallback string) string {
	if len(raw) == 0 || string(raw) == "null" {
		return fallback
	}

	var text string
	if err := json.Unmarshal(raw, &text); err == nil {
		return text
	}
	return rawScalarString(raw, fallback)
}

func rawScalarString(raw json.RawMessage, fallback string) string {
	if len(raw) == 0 {
		return fallback
	}

	trimmed := strings.TrimSpace(string(raw))
	if trimmed == "null" {
		return "None"
	}
	if trimmed == "true" {
		return "True"
	}
	if trimmed == "false" {
		return "False"
	}

	var text string
	if err := json.Unmarshal(raw, &text); err == nil {
		return text
	}
	return trimmed
}

func rawNumber(raw json.RawMessage, fallback float64) float64 {
	if len(raw) == 0 {
		return fallback
	}
	trimmed := strings.TrimSpace(string(raw))
	if trimmed == "true" {
		return 1
	}
	if trimmed == "false" {
		return 0
	}
	number, err := strconv.ParseFloat(trimmed, 64)
	if err != nil {
		return fallback
	}
	return number
}
